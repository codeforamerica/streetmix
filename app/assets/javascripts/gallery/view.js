var THUMBNAIL_WIDTH = 180;
var THUMBNAIL_HEIGHT = 110;
var THUMBNAIL_MULTIPLIER = .1 * 2;
var BACKGROUND_DIRT_COLOUR = 'rgb(53, 45, 39)';

var TRACK_ACTION_OPEN_GALLERY = 'Open gallery';

var galleryVisible = false;

var galleryTwitterId = null;
var galleryStreetId = null;
var galleryStreetLoaded = false;

function _repeatReceiveGalleryData() {
  _loadGalleryContents();
}

function _updateGallerySelection() {
  var els = document.querySelectorAll('#gallery .streets .selected');
  for (var i = 0, el; el = els[i]; i++) {
    el.classList.remove('selected');
  }

  var el = document.querySelector('#gallery .streets [streetId="' +
      galleryStreetId + '"]');
  if (el) {
    el.classList.add('selected');
  }
}

function _switchGalleryStreet(id) {
  galleryStreetId = id;

  _updateGallerySelection();
  _fetchGalleryStreet(galleryStreetId);
}

function _onGalleryStreetClick(event) {
  if (event.shiftKey || event.ctrlKey || event.metaKey) {
    return;
  }

  var el = this;
  _switchGalleryStreet(el.getAttribute('streetId'));

  event.preventDefault();
}

function _updateGalleryStreetCount() {
  if (galleryTwitterId) {
    var streetCount = document.querySelectorAll('#gallery .streets li').length;

    switch (streetCount) {
      case 0:
        var text = msg('STREET_COUNT_0');
        break;
      case 1:
        var text = msg('STREET_COUNT_1');
        break;
      default:
        var text = msg('STREET_COUNT_MANY', { streetCount: streetCount });
        break;
    }
  } else {
    var text = '';
  }
  document.querySelector('#gallery .street-count').innerHTML = text;
}

function _receiveGalleryData(transmission) {
  document.querySelector('#gallery .loading').classList.remove('visible');

  for (var i in transmission.streets) {
    var galleryStreet = transmission.streets[i];

    // There is a bug where sometimes street data is non-existent for an unknown reason
    // Skip over so that the rest of gallery will display
    if (!galleryStreet.data) continue;

    _updateToLatestSchemaVersion(galleryStreet.data.street);

    var el = document.createElement('li');

    var anchorEl = document.createElement('a');

    /*if (!galleryTwitterId && (galleryStreet.data.undoStack.length <= 4)) {
      anchorEl.classList.add('virgin');
    }*/

    galleryStreet.creatorId =
        (galleryStreet.creator && galleryStreet.creator.twitterId);

    galleryStreet.name = galleryStreet.name || DEFAULT_NAME;

    anchorEl.href = _getStreetUrl(galleryStreet);

    anchorEl.streetName = galleryStreet.name;
    anchorEl.setAttribute('streetId', galleryStreet.id);

    if (galleryStreetId == galleryStreet.id) {
      anchorEl.classList.add('selected');
    }

    $(anchorEl).click(_onGalleryStreetClick);

    var thumbnailEl = document.createElement('canvas');
    thumbnailEl.width = THUMBNAIL_WIDTH * system.hiDpi * 2;
    thumbnailEl.height = THUMBNAIL_HEIGHT * system.hiDpi * 2;
    var ctx = thumbnailEl.getContext('2d');
    _drawStreetThumbnail(ctx, galleryStreet.data.street,
        THUMBNAIL_WIDTH * 2, THUMBNAIL_HEIGHT * 2, THUMBNAIL_MULTIPLIER, true, false, true, false, false);
    anchorEl.appendChild(thumbnailEl);

    var nameEl = document.createElement('div');
    nameEl.classList.add('street-name');
    nameEl.innerHTML = '<div></div>';
    anchorEl.appendChild(nameEl);

    $(nameEl.querySelector('div')).text(galleryStreet.name);
    _updateStreetNameFont(nameEl);

    var date = moment(galleryStreet.updatedAt);
    var dateEl = document.createElement('span');
    dateEl.classList.add('date');
    dateEl.innerHTML = _formatDate(date);
    anchorEl.appendChild(dateEl);

    if (!galleryTwitterId) {
      var creatorEl = document.createElement('span');
      creatorEl.classList.add('creator');

      var creatorName = galleryStreet.creatorId || msg('USER_ANONYMOUS');

      creatorEl.innerHTML = creatorName;
      anchorEl.appendChild(creatorEl);
    }

    // Only show delete links if you own the street
    if (signedIn && (galleryStreet.creatorId == signInData.twitterId)) {
      var removeEl = document.createElement('button');
      removeEl.classList.add('remove');
      removeEl.addEventListener('click', _onDeleteGalleryStreet);
      removeEl.innerHTML = msg('UI_GLYPH_X');
      removeEl.title = msg('TOOLTIP_DELETE_STREET');
      anchorEl.appendChild(removeEl);
    }

    el.appendChild(anchorEl);
    document.querySelector('#gallery .streets').appendChild(el);
  }

  var streetCount = document.querySelectorAll('#gallery .streets li').length;

  if (((mode == MODES.USER_GALLERY) && streetCount) || (mode == MODES.GLOBAL_GALLERY)) {
    _switchGalleryStreet(transmission.streets[0].id);
  }

  var el = document.querySelector('#gallery .selected');
  if (el) {
    el.scrollIntoView();
    document.querySelector('#gallery').scrollTop = 0;
  }

  _updateScrollButtons();

  _updateGalleryStreetCount();
}

function _loadGalleryContents() {
  var els = document.querySelectorAll('#gallery .streets li');
  for (var i = 0, el; el = els[i]; i++) {
    _removeElFromDom(el);
  }

  //document.querySelector('#gallery .streets').innerHTML = '';
  document.querySelector('#gallery .loading').classList.add('visible');
  document.querySelector('#gallery .error-loading').classList.remove('visible');

  _fetchGalleryData();
}

function _showGallery(twitterId, instant, signInPromo) {
  if (readOnly) {
    return;
  }

  Stmx.app.eventTracking.track(TRACK_CATEGORY_INTERACTION, TRACK_ACTION_OPEN_GALLERY, userId, null, false);

  galleryVisible = true;
  galleryStreetLoaded = true;
  galleryStreetId = street.id;
  galleryTwitterId = twitterId;

  if (signInPromo) {

  } else {
    if (twitterId) {
      document.querySelector('#gallery .avatar').setAttribute('twitterId', galleryTwitterId);
      document.querySelector('#gallery .avatar').removeAttribute('loaded');
      _fetchAvatars();
      document.querySelector('#gallery .user-id').innerHTML = galleryTwitterId;

      var linkEl = document.createElement('a');
      // TODO const
      linkEl.href = 'https://twitter.com/' + galleryTwitterId;
      linkEl.innerHTML = 'Twitter profile Â»';
      linkEl.classList.add('twitter-profile');
      linkEl.target = '_blank';
      document.querySelector('#gallery .user-id').appendChild(linkEl);

    } else {
      document.querySelector('#gallery .user-id').innerHTML = 'All streets';
    }


    document.querySelector('#gallery .street-count').innerHTML = '';

    // TODO no class, but type?
    if (!twitterId) {
      document.querySelector('#gallery').classList.add('all-streets');
      document.querySelector('#gallery').classList.remove('another-user');
    } else if (signedIn && (twitterId == signInData.twitterId)) {
      document.querySelector('#gallery').classList.remove('another-user');
      document.querySelector('#gallery').classList.remove('all-streets');
    } else {
      document.querySelector('#gallery').classList.add('another-user');
      document.querySelector('#gallery').classList.remove('all-streets');
    }
  }

  _hideControls();
  _statusMessage.hide();
  document.querySelector('#gallery .sign-in-promo').classList.remove('visible');

  if (instant) {
    document.body.classList.add('gallery-no-move-transition');
  }
  document.body.classList.add('gallery-visible');

  if (instant) {
    window.setTimeout(function() {
      document.body.classList.remove('gallery-no-move-transition');
    }, 0);
  }

  if ((mode == MODES.USER_GALLERY) || (mode == MODES.GLOBAL_GALLERY)) {
    // Prevents showing old street before the proper street loads
    _showError(ERRORS.NO_STREET, false);
  }

  if (!signInPromo) {
    _loadGalleryContents();
    _updatePageUrl(true);
  } else {
    document.querySelector('#gallery .sign-in-promo').classList.add('visible');
  }
}

function _onGalleryShieldClick(event) {
  _hideGallery(false);
}

function _hideGallery(instant) {
  if ((currentErrorType != ERRORS.NO_STREET) && galleryStreetLoaded) {
    galleryVisible = false;

    if (instant) {
      document.body.classList.add('gallery-no-move-transition');
    }
    document.body.classList.remove('gallery-visible');

    if (instant) {
      window.setTimeout(function() {
        document.body.classList.remove('gallery-no-move-transition');
      }, 0);
    }

    _onWindowFocus();

    if (!abortEverything) {
      _updatePageUrl();
    }

    mode = MODES.CONTINUE;
  }
}

function _updateGalleryShield() {
  document.querySelector('#gallery-shield').style.width = 0;
  window.setTimeout(function() {
    document.querySelector('#gallery-shield').style.height =
        system.viewportHeight + 'px';
    document.querySelector('#gallery-shield').style.width =
        document.querySelector('#street-section-outer').scrollWidth + 'px';
  }, 0);
}


function _onDeleteGalleryStreet(event) {
  var el = event.target.parentNode;
  var name = el.streetName;

  _ignoreWindowFocusMomentarily();
  // TODO escape name
  if (confirm(msg('PROMPT_DELETE_STREET', { name: name }))) {
    if (el.getAttribute('streetId') == street.id) {
      _showError(ERRORS.NO_STREET, false);
    }

    _sendDeleteStreetToServer(el.getAttribute('streetId'));

    _removeElFromDom(el.parentNode);
    _updateGalleryStreetCount();
  }

  event.preventDefault();
  event.stopPropagation();
}

function _onMyStreetsClick(event) {
  if (event.shiftKey || event.ctrlKey || event.metaKey) {
    return;
  }

  if (signedIn) {
    _showGallery(signInData.twitterId, false);
  } else {
    _showGallery(false, false, true);
  }

  event.preventDefault();
}
