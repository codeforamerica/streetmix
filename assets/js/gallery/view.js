var THUMBNAIL_WIDTH = 180
var THUMBNAIL_HEIGHT = 110
var THUMBNAIL_MULTIPLIER = .1 * 2
var BACKGROUND_DIRT_COLOUR = 'rgb(53, 45, 39)'

var TRACK_ACTION_OPEN_GALLERY = 'Open gallery'

var galleryVisible = false

var galleryUserId = null
var galleryStreetId = null
var galleryStreetLoaded = false

// set to true when the current street is deleted from the gallery
// this prevents the gallery from being hidden while no street is shown
var galleryNoStreetSelected = false

function _initGallery () {
  document.querySelector('#new-street').href = URL_NEW_STREET
  document.querySelector('#copy-last-street').href = URL_NEW_STREET_COPY_LAST
}

function _repeatReceiveGalleryData () {
  _loadGalleryContents()
}

function _updateGallerySelection () {
  var els = document.querySelectorAll('#gallery .streets .selected')
  for (var i = 0, el; el = els[i]; i++) {
    el.classList.remove('selected')
  }

  var el = document.querySelector('#gallery .streets [streetId="' +
    galleryStreetId + '"]')
  if (el) {
    el.classList.add('selected')
  }
}

function _switchGalleryStreet (id) {
  galleryStreetId = id
  galleryNoStreetSelected = false

  _updateGallerySelection()
  _fetchGalleryStreet(galleryStreetId)
}

function _onGalleryStreetClick (event) {
  if (event.shiftKey || event.ctrlKey || event.metaKey) {
    return
  }

  var el = this
  _switchGalleryStreet(el.getAttribute('streetId'))

  event.preventDefault()
}

function _updateGalleryStreetCount () {
  if (galleryUserId) {
    var streetCount = document.querySelectorAll('#gallery .streets li').length

    switch (streetCount) {
      case 0:
        var text = msg('STREET_COUNT_0')
        break
      case 1:
        var text = msg('STREET_COUNT_1')
        break
      default:
        var text = msg('STREET_COUNT_MANY', { streetCount: streetCount })
        break
    }
  } else {
    var text = ''
  }
  document.querySelector('#gallery .street-count').innerHTML = text
}

function _receiveGalleryData (transmission) {
  document.querySelector('#gallery .loading').classList.remove('visible')

  for (var i in transmission.streets) {
    var galleryStreet = transmission.streets[i]

    // There is a bug where sometimes street data is non-existent for an unknown reason
    // Skip over so that the rest of gallery will display
    if (!galleryStreet.data) continue

    _updateToLatestSchemaVersion(galleryStreet.data.street)

    var el = document.createElement('li')

    var anchorEl = document.createElement('a')

    /*if (!galleryUserId && (galleryStreet.data.undoStack.length <= 4)) {
      anchorEl.classList.add('virgin')
    }*/

    galleryStreet.creatorId =
      (galleryStreet.creator && galleryStreet.creator.id)

    galleryStreet.name = galleryStreet.name || DEFAULT_NAME

    anchorEl.href = _getStreetUrl(galleryStreet)

    anchorEl.streetName = galleryStreet.name
    anchorEl.setAttribute('streetId', galleryStreet.id)

    if (galleryStreetId == galleryStreet.id) {
      anchorEl.classList.add('selected')
    }

    anchorEl.addEventListener('click', _onGalleryStreetClick)

    var thumbnailEl = document.createElement('canvas')
    thumbnailEl.width = THUMBNAIL_WIDTH * system.hiDpi * 2
    thumbnailEl.height = THUMBNAIL_HEIGHT * system.hiDpi * 2
    var ctx = thumbnailEl.getContext('2d')
    drawStreetThumbnail(ctx, galleryStreet.data.street,
      THUMBNAIL_WIDTH * 2, THUMBNAIL_HEIGHT * 2, THUMBNAIL_MULTIPLIER, true, false, true, false, false)
    anchorEl.appendChild(thumbnailEl)

    var nameEl = document.createElement('div')
    nameEl.className = 'street-name'
    anchorEl.appendChild(nameEl)

    var streetName = new StreetName(nameEl, galleryStreet.name)

    var dateEl = document.createElement('span')
    dateEl.classList.add('date')
    dateEl.innerHTML = _formatDate(galleryStreet.updatedAt)
    anchorEl.appendChild(dateEl)

    if (!galleryUserId) {
      var creatorEl = document.createElement('span')
      creatorEl.classList.add('creator')

      var creatorName = galleryStreet.creatorId || msg('USER_ANONYMOUS')

      creatorEl.innerHTML = creatorName
      anchorEl.appendChild(creatorEl)
    }

    // Only show delete links if you own the street
    if (signedIn && (galleryStreet.creatorId == signInData.userId)) {
      var removeEl = document.createElement('button')
      removeEl.classList.add('remove')
      removeEl.addEventListener('pointerdown', _onDeleteGalleryStreet)
      removeEl.innerHTML = msg('UI_GLYPH_X')
      removeEl.title = msg('TOOLTIP_DELETE_STREET')
      anchorEl.appendChild(removeEl)
    }

    el.appendChild(anchorEl)
    document.querySelector('#gallery .streets').appendChild(el)
  }

  var streetCount = document.querySelectorAll('#gallery .streets li').length

  if (((mode == MODES.USER_GALLERY) && streetCount) || (mode == MODES.GLOBAL_GALLERY)) {
    _switchGalleryStreet(transmission.streets[0].id)
  }

  var el = document.querySelector('#gallery .selected')
  if (el) {
    el.scrollIntoView()
    document.querySelector('#gallery').scrollTop = 0
  }

  _updateScrollButtons()

  _updateGalleryStreetCount()
}

function _loadGalleryContents () {
  var els = document.querySelectorAll('#gallery .streets li')
  for (var i = 0, el; el = els[i]; i++) {
    removeElFromDOM(el)
  }

  // document.querySelector('#gallery .streets').innerHTML = ''
  document.querySelector('#gallery .loading').classList.add('visible')
  document.querySelector('#gallery .error-loading').classList.remove('visible')

  _fetchGalleryData()
}

function _showGallery (userId, instant, signInPromo) {
  if (app.readOnly) {
    return
  }

  trackEvent('Interaction', TRACK_ACTION_OPEN_GALLERY,
    userId, null, false)

  galleryVisible = true
  galleryStreetLoaded = true
  galleryStreetId = street.id
  galleryUserId = userId

  if (signInPromo) {
  } else {
    if (userId) {
      document.querySelector('#gallery .avatar').setAttribute('userId', galleryUserId)
      document.querySelector('#gallery .avatar').removeAttribute('loaded')
      _fetchAvatars()
      document.querySelector('#gallery .user-id').innerHTML = galleryUserId

      var linkEl = document.createElement('a')
      // TODO const
      linkEl.href = 'https://twitter.com/' + galleryUserId
      linkEl.innerHTML = 'Twitter profile Â»'
      linkEl.classList.add('twitter-profile')
      linkEl.target = '_blank'
      document.querySelector('#gallery .user-id').appendChild(linkEl)

    } else {
      document.querySelector('#gallery .user-id').innerHTML = 'All streets'
    }

    document.querySelector('#gallery .street-count').innerHTML = ''

    // TODO no class, but type?
    if (!userId) {
      document.querySelector('#gallery').classList.add('all-streets')
      document.querySelector('#gallery').classList.remove('another-user')
    } else if (signedIn && (userId == signInData.userId)) {
      document.querySelector('#gallery').classList.remove('another-user')
      document.querySelector('#gallery').classList.remove('all-streets')
    } else {
      document.querySelector('#gallery').classList.add('another-user')
      document.querySelector('#gallery').classList.remove('all-streets')
    }
  }

  _hideControls()
  hideStatusMessage()
  document.querySelector('#gallery .sign-in-promo').classList.remove('visible')

  if (instant) {
    document.body.classList.add('gallery-no-move-transition')
  }
  document.body.classList.add('gallery-visible')

  if (instant) {
    window.setTimeout(function () {
      document.body.classList.remove('gallery-no-move-transition')
    }, 0)
  }

  if ((mode == MODES.USER_GALLERY) || (mode == MODES.GLOBAL_GALLERY)) {
    // Prevents showing old street before the proper street loads
    _showError(ERRORS.NO_STREET, false)
  }

  if (!signInPromo) {
    _loadGalleryContents()
    _updatePageUrl(true)
  } else {
    document.querySelector('#gallery .sign-in-promo').classList.add('visible')
  }
}

function _onGalleryShieldClick (event) {
  _hideGallery(false)
}

function _hideGallery (instant) {
  // Do not hide the gallery if there is no street selected.
  if (galleryNoStreetSelected === true) {
    return
  }

  if (galleryStreetLoaded) {
    galleryVisible = false

    if (instant) {
      document.body.classList.add('gallery-no-move-transition')
    }
    document.body.classList.remove('gallery-visible')

    if (instant) {
      window.setTimeout(function () {
        document.body.classList.remove('gallery-no-move-transition')
      }, 0)
    }

    _onWindowFocus()

    if (!abortEverything) {
      _updatePageUrl()
    }

    mode = MODES.CONTINUE
  }
}

function _updateGalleryShield () {
  document.querySelector('#gallery-shield').style.width = 0
  window.setTimeout(function () {
    document.querySelector('#gallery-shield').style.height =
      system.viewportHeight + 'px'
    document.querySelector('#gallery-shield').style.width =
      document.querySelector('#street-section-outer').scrollWidth + 'px'
  }, 0)
}

function _onDeleteGalleryStreet (event) {
  var el = event.target.parentNode
  var name = el.streetName

  // TODO escape name
  if (confirm(msg('PROMPT_DELETE_STREET', { name: name }))) {
    if (el.getAttribute('streetId') == street.id) {
      galleryNoStreetSelected = true
      _showError(ERRORS.NO_STREET, false)
    }

    _sendDeleteStreetToServer(el.getAttribute('streetId'))

    removeElFromDOM(el.parentNode)
    _updateGalleryStreetCount()
  }

  event.preventDefault()
  event.stopPropagation()
}

function _onMyStreetsClick (event) {
  if (event.shiftKey || event.ctrlKey || event.metaKey) {
    return
  }

  if (signedIn) {
    _showGallery(signInData.userId, false)
  } else {
    _showGallery(false, false, true)
  }

  event.preventDefault()
}
