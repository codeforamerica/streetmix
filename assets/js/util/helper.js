function RandomGenerator() {
  this.randSeed = 0;
}

RandomGenerator.prototype.rand = function() {
  var t32 = 0x100000000;
  var constant = 134775813;
  var x = (constant * this.randSeed + 1);
  return (this.randSeed = x % t32) / t32;
}

RandomGenerator.prototype.seed = function(seed) {
  this.randSeed = seed;
}

function htmlEncode(value){
  //create a in-memory div, set it's inner text(which jQuery automatically encodes)
  //then grab the encoded contents back out.  The div never exists on the page.
  return $('<div/>').text(value).html();
}


function msg(messageId, data) {
  if (data) {
    return MESSAGES[messageId].supplant(data);
  } else {
    return MESSAGES[messageId];
  }
}

String.prototype.supplant = function(o) {
  return this.replace(/\[\[([^\[\]]*)\]\]/g,
    function (a, b) {
      var r = o[b];
      return typeof r === 'string' || typeof r === 'number' ? r : a;
    }
  );
};

function _createTimeout(fn, data, delay) {
  window.setTimeout(function() { fn.call(null, data); }, delay);
}

function _removeElFromDom(el) {
  if (el && el.parentNode) {
    el.parentNode.removeChild(el);
  }
}

function _getElAbsolutePos(el) {
  var pos = [0, 0];

  do {
    pos[0] += el.offsetLeft + (el.cssTransformLeft || 0);
    pos[1] += el.offsetTop + (el.cssTransformTop || 0);

    el = el.offsetParent;
  } while (el);

  return pos;
}

function _clone(obj) {
  if ($.isArray(obj)) {
    return $.extend(true, [], obj);
  } else {
    return $.extend(true, {}, obj);
  }
}
