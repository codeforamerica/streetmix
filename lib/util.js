exports.parseLoginToken = function (req) {
  // authHeader is a string that looks like this:
  // `Streetmix realm="" loginToken="xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxx" userId="user"`
  const authHeader = req.header('Authorization')

  if (!authHeader) {
    return null
  }

  const auth = authHeader.split(/\s+/)

  if (auth.length < 4) {
    return null
  }

  const authScheme = auth.shift()

  // Convert the auth string to an object
  const authObject = auth.reduce((accumulator, item) => {
    const [ key, value ] = item.split('=')
    accumulator[key] = value.replace(/(^"|"$)/g, '')
    return accumulator
  }, {})

  if ((authScheme !== 'Streetmix') || (authObject.realm !== '')) {
    return null
  }

  return authObject
}

exports.ERRORS = {
  USER_NOT_FOUND: 'USER_NOT_FOUND',
  INTERNAL_ERROR: 'INTERNAL_ERROR',
  STREET_NOT_FOUND: 'STREET_NOT_FOUND',
  STREET_DELETED: 'STREET_DELETED',
  UNAUTHORISED_ACCESS: 'UNAUTHORISED_ACCESS',
  FORBIDDEN_REQUEST: 'FORBIDDEN_REQUEST',
  CANNOT_CREATE_STREET: 'CANNOT_CREATE_STREET',
  CANNOT_UPDATE_STREET: 'CANNOT_UPDATE_STREET',
  CANNOT_GET_STREET: 'CANNOT_GET_STREET',
  CANNOT_GET_USER: 'CANNOT_GET_USER'
}

exports.SAVE_THUMBNAIL_EVENTS = {
  INITIAL: 'INITIAL',
  SHARE: 'SHARE',
  TIMER: 'TIMER',
  BEFOREUNLOAD: 'BEFOREUNLOAD',
  PREVIOUS_STREET: 'PREVIOUS_STREET',
  TEST: 'TEST'
}

exports.requestIp = function (req) {
  if (req.headers['x-forwarded-for'] !== undefined) {
    return req.headers['x-forwarded-for'].split(', ')[0]
  } else {
    return req.connection.remoteAddress
  }
}
