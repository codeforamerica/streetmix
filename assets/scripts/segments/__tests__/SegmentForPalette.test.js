/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import { SegmentForPalette } from '../SegmentForPalette'
import { getSegmentInfo, getSegmentVariantInfo } from '../info'
import { getVariantInfoDimensions } from '../view'
import { generateRandSeed } from '../../util/random'
import { IntlProvider } from 'react-intl'

jest.mock('../../streets/data_model', () => {})
jest.mock('../info')
jest.mock('../view')
jest.mock('../../util/random')

function connectDropTarget (el) { return el }
function connectDragSource (el) { return el }

describe('SegmentForPalette', () => {
  const connectDragPreview = jest.fn()
  const intlProvider = new IntlProvider({ locale: 'en' }, {})
  const { intl } = intlProvider.getChildContext()
  beforeEach(() => {
    const dimensions = { left: 100, right: 200 }
    generateRandSeed.mockImplementation(() => 42)
    getVariantInfoDimensions.mockImplementation(() => dimensions)
  })
  describe('on mount', () => {
    it('connects to drag', () => {
      shallow(<SegmentForPalette connectDropTarget={connectDropTarget} connectDragSource={connectDragSource} connectDragPreview={connectDragPreview} type={''} variantString={''} intl={intl} />)
      expect(connectDragPreview).toHaveBeenCalledTimes(1)
    })
  })
  it('renders width correctly depending on the dimension', () => {
    const wrapper = shallow(<SegmentForPalette connectDropTarget={connectDropTarget} connectDragSource={connectDragSource} connectDragPreview={connectDragPreview} type={''} variantString={''} intl={intl} />)
    expect(wrapper).toMatchSnapshot()
  })
  describe('mouseover', () => {
    it('calls onPointerOver with Variant name', () => {
      const onPointerOver = jest.fn()
      const event = { target: { getBoundingClientRect: () => 1 } }
      const variant = { name: 'Variant' }
      const segment = { nameKey: 'key' }
      getSegmentInfo.mockImplementation(() => segment)
      getSegmentVariantInfo.mockImplementation(() => variant)
      const wrapper = shallow(<SegmentForPalette connectDropTarget={connectDropTarget} connectDragSource={connectDragSource} connectDragPreview={connectDragPreview} type={''} variantString={''} intl={intl} onPointerOver={onPointerOver} />)
      wrapper.simulate('pointerover', event)
      expect(onPointerOver).toHaveBeenCalledTimes(1)
      expect(onPointerOver).toHaveBeenCalledWith(event, 'Variant', 1)
    })
    it('calls onPointerOver with Segment name', () => {
      const onPointerOver = jest.fn()
      const event = { target: { getBoundingClientRect: () => 1 } }
      const variant = { }
      const segment = { name: 'Segment', nameKey: 'key' }
      getSegmentInfo.mockImplementation(() => segment)
      getSegmentVariantInfo.mockImplementation(() => variant)
      const wrapper = shallow(<SegmentForPalette connectDropTarget={connectDropTarget} connectDragSource={connectDragSource} connectDragPreview={connectDragPreview} type={''} variantString={''} intl={intl} onPointerOver={onPointerOver} />)
      wrapper.simulate('pointerover', event)
      expect(onPointerOver).toHaveBeenCalledTimes(1)
      expect(onPointerOver).toHaveBeenCalledWith(event, 'Segment', 1)
    })
  })
})
