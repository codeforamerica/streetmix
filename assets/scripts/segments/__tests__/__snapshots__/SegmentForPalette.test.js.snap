// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SegmentForPalette renders width correctly depending on the dimension 1`] = `
<DocumentFragment>
  <li
    class="segment segment-in-palette"
    data-testid="segment-for-palette"
    style="width: 424px;"
  >
    <div
      style="height: 80px;"
    >
      <canvas
        class="segment-image"
        height="445"
        style="width: 424px; height: 445px; left: 400px;"
        width="424"
      />
    </div>
  </li>
</DocumentFragment>
`;
